rcon/routines.py

```diff
@@ -56,21 +41,56 @@ def toggle_votekick(rcon: RecordedRcon):
         rcon.set_votekick_enabled(True)
 
 
-def run():
+async def scheduled_routines(rcon: RecordedRcon):
     max_fails = 5
-    from rcon.settings import SERVER_INFO
-
-    rcon = RecordedRcon(SERVER_INFO)
-    # recorder = MapsRecorder(rcon)
-
     while True:
         try:
-            # recorder.detect_map_change()
             toggle_votekick(rcon)
             VoteMap().vote_map_reminder(rcon)
         except CommandFailedError:
             max_fails -= 1
             if max_fails <= 0:
-                logger.exception("Routines 5 failures in a row. Stopping")
+                logger.exception("scheduled routines 5 failures in a row. Stopping")
                 raise
-        time.sleep(30)
+        await asyncio.sleep(30)
+
+
+async def auto_broadcast(log: logging.Logger, rcon: RecordedRcon):
+    log.debug("Starting broadcasts")
+    await broadcast.run(log, rcon)
+
+
+async def expire_vips(log: logging.Logger, rcon: RecordedRcon):
+    log.debug("Starting expiring vips")
+    await service.run(rcon)
+
+
+def component_logger(name: str, server_number: str) -> logging.Logger:
+    filename = f"{name}_{server_number}.log"
+    handler = logging.FileHandler(filename)
+
+    log = logging.getLogger(name)
+    level = os.getenv("LOGGING_LEVEL")
+    if level is not None:
+        log.setLevel(level)
+    log.addHandler(handler)
+
+    return log
+
+
+async def run():
+    from rcon.settings import SERVER_INFO
+    rcon = RecordedRcon(SERVER_INFO)
+    server_number = os.getenv("SERVER_NUMBER")
+
+    tasks = asyncio.gather(
+        scheduled_routines(rcon),
+        auto_broadcast(component_logger("broadcasts", server_number), rcon),
+        expire_vips(component_logger("expiring_vips", server_number), rcon),
```

Comment:
Hmm is that going to work? I'm afraid the broadcast will be quite out of sync especially if ppl use it with 1 second messages. Since those are not really async function, when they do they blocking IO they won't be releasing their time to other function, only when they reach an async sleep. so the whole timing will be highly dependent on how long the preceding function take to run. Could be quite messy no?

Reply:
When there is quite some traffic in these things (routines, auto-broadcasts and expire_vips), the execution might be delayed a bit, yes. However, is 1-second delayed messages in auto-broadcast really a thing that we see in the wild or even want to really support? 🤔 
If so, this whole thing would benefit from the RCon connections being truly async, yes.